Non-visible symbol in wildcard import taking precedence over visible symbol in current package

## Reproduction steps

Scala version: 2.13.10

https://github.com/armanbilge/sandbox/commit/0c6533cff815c823c13dd7397289ebc49edf0526

## Problem

```
sbt:sandbox> bar/compile
[info] compiling 1 Scala source to /workspace/sandbox/foo/target/scala-2.13/classes ...
[info] compiling 2 Scala sources to /workspace/sandbox/bar/target/scala-2.13/classes ...
[error] /workspace/sandbox/bar/src/main/scala/bar.scala:7:3: object Foo in package foo cannot be accessed as a member of package foo from object Bar in package baz
[error]   Foo
[error]   ^
[error] one error found
[error] (bar / Compile / compileIncremental) Compilation failed
[error] Total time: 0 s, completed Feb 22, 2023, 7:15:12 AM
sbt:sandbox> 
```

Meanwhile, there is a visible `Foo` in the current package, which is the one we want.

It's the separate compilation bug where it doesn't do "package private" correctly.
```
➜  t12736 scalac -d /tmp foo.scala
➜  t12736 scalac -d /tmp -cp /tmp b*.scala
baz.scala:8: error: object Foo in package foo cannot be accessed as a member of package foo from object Bar in package baz
  def test() = println(Foo)
                       ^
1 error
```
Not sure I knew it manifested this way, but it rings a bell.

--- a/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
@@ -1246,9 +1246,11 @@ trait Contexts { self: Analyzer =>
 
     /** If the given import is permitted, fetch the symbol and filter for accessibility.
      */
-    private[Contexts] def importedAccessibleSymbol(imp: ImportInfo, sym: => Symbol): Symbol =
+    private[Contexts] def importedAccessibleSymbol(imp: ImportInfo, sym: => Symbol): Symbol = {
       if (isExcludedRootImport(imp)) NoSymbol
-      else sym.filter(isAccessible(_, imp.qual.tpe, superAccess = false))
+      else sym.filter(s => s.exists && isAccessible(s, imp.qual.tpe, superAccess = false))
+      // `exists` above completes SymbolLoaders, which sets the symbol's access flags (scala/bug#12736)
+    }
 
     private def isExcludedRootImport(imp: ImportInfo): Boolean =
       imp.isRootImport && excludedRootImportsCached.get(unit).exists(_.contains(imp.qual.symbol))
@@ -1508,6 +1510,9 @@ trait Contexts { self: Analyzer =>
       val importCursor = new ImportCursor(thisContext, name)
       import importCursor.{imp1, imp2}
 
+      // The symbol resolved by the given import for `name`, paired with the selector that was used.
+      // If `requireExplicit`, then only "named" or "specific" selectors are considered.
+      // In addition, the symbol must be accessible (in the current context) and satisfy the `qualifies` predicate.
       def lookupImport(imp: ImportInfo, requireExplicit: Boolean): (ImportSelector, Symbol) = {
         val pair @ (sel, sym) = imp.importedSelectedSymbol(name, requireExplicit)
         if (sym == NoSymbol) pair

--- a/src/compiler/scala/tools/nsc/typechecker/Infer.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Infer.scala
@@ -286,7 +286,7 @@ trait Infer extends Checkable {
         ErrorUtils.issueTypeError(error)(context)
         ErrorType
       }
-      def accessible = sym filter (alt => context.isAccessible(alt, pre, site.isInstanceOf[Super])) match {
+      def accessible = sym.filter(context.isAccessible(_, pre, site.isInstanceOf[Super])) match {
         case NoSymbol if sym.isJavaDefined && context.unit.isJava => sym  // don't try to second guess Java; see #4402
         case sym1                                                 => sym1
       }

--- a/src/reflect/scala/reflect/internal/Symbols.scala
+++ b/src/reflect/scala/reflect/internal/Symbols.scala
@@ -1095,21 +1095,21 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
       }
 
     def exists: Boolean = !isTopLevel || {
-      val isSourceLoader = rawInfo match {
-        case sl: SymLoader => sl.fromSource
-        case _             => false
-      }
-      def warnIfSourceLoader(): Unit = {
+      def warnIfSourceLoader(): false = {
+        val isSourceLoader = rawInfo match {
+          case sl: SymLoader => sl.fromSource
+          case _             => false
+        }
+        // Predef is completed early due to its autoimport; we used to get here when type checking its
+        // parent LowPriorityImplicits. See comment in c5441dc for more elaboration.
+        // Since the fix for scala/bug#7335 Predef parents must be defined in Predef.scala, and we should not
+        // get here anymore.
         if (isSourceLoader)
-          // Predef is completed early due to its autoimport; we used to get here when type checking its
-          // parent LowPriorityImplicits. See comment in c5441dc for more elaboration.
-          // Since the fix for scala/bug#7335 Predef parents must be defined in Predef.scala, and we should not
-          // get here anymore.
           devWarning(s"calling Symbol#exists with sourcefile based symbol loader may give incorrect results.")
+        false
       }
-
-      rawInfo load this
-      rawInfo != NoType || { warnIfSourceLoader(); false }
+      rawInfo.load(this)
+      rawInfo != NoType || warnIfSourceLoader()
     }
 
     final def isInitialized: Boolean =

--- a/test/files/pos/t12736/bar_2.scala
+++ b/test/files/pos/t12736/bar_2.scala
@@ -1,0 +1,3 @@
+package bar
+
+private object Foo

--- a/test/files/pos/t12736/baz_2.scala
+++ b/test/files/pos/t12736/baz_2.scala
@@ -1,0 +1,15 @@
+package bar
+package baz
+
+import foo._
+
+object Bar {
+  def test() = println(Foo)
+}
+
+/*
+baz_2.scala:8: error: object Foo in package foo cannot be accessed as a member of package foo from object Bar in package baz
+  def test() = println(Foo)
+                       ^
+1 error
+*/

--- a/test/files/pos/t12736/foo_1.scala
+++ b/test/files/pos/t12736/foo_1.scala
@@ -1,0 +1,3 @@
+package foo
+
+private object Foo

